Add support to parsing lts/* aliases in legacy config files

augustobmoura opened the pull request:
Partially solves #193, we still cannot run `asdf install nodejs "lts/*"`, but now it translates the .nvmrc version correctly from `lts/codename` to `lts-codename` and `lts/*` to `lts`

dcalhoun commented:
@augustobmoura thank you for implementing this. I'm excited to use it!

dcalhoun commented:
@augustobmoura I switched to your branch to test your proposed changes. Unfortunately, I don't think the changes to the `parse-legacy-file` hook alone will accomplish the end goal of translating an alias found within a `.nvmrc` file to a valid Node.js version installed via `asdf`.

<img width="737" alt="asdf-nodejs-lts-alias" src="https://user-images.githubusercontent.com/438664/105720205-cb3b0b00-5ee8-11eb-8be2-3037d9d822e3.png">

As you can see in my example image, `asdf` is unable to translate the `.nvmrc` alias of `lts/*` to the matching valid `14.15.4` (current latest LTS) version of Node.js that is already installed. The expectation would be that `asdf` recognizes that the "LTS" is already installed. However, it currently still requests that `asdf install nodejs lts` be run instead.

From your perspective, do you think this type of alias is achievable in some form?

augustobmoura commented:
> <img alt="asdf-nodejs-lts-alias" width="737" src="https://user-images.githubusercontent.com/438664/105720205-cb3b0b00-5ee8-11eb-8be2-3037d9d822e3.png">

From the image it looks like you ran in another error, after `asdf install nodejs lts` the command `node -v` should've correctly point to 14.15.4 or the last version for that lts. You ran into the error `local: -A invalid option` during the install, can you verify if you are pointing to the right branch from my repo? I have some work of other pull requests in my master

To install a specific branch you need to manually checkout it, you can run the following to install the branch:

``` bash
asdf plugin add nodejs https://github.com/augustobmoura/asdf-nodejs
git -C "${ASDF_DIR:-$HOME/.asdf}/plugins/nodejs" checkout feature/nvmrc-lts-aliases
```

Can you please also send your bash version and OS? It might help with my other work

augustobmoura commented:
Also can you verify if is not related to #199?

dcalhoun commented:
> From the image it looks like you ran in another error, after `asdf install nodejs lts`

@augustobmoura no, the end of my image was the last command I ran. It was showcasing that `asdf install nodejs lts` resulted in a failure.

> the command `node -v` should've correctly point to 14.15.4 or the last version for that lts. You ran into the error `local: -A invalid option` during the install

Correct, I am attempting to showcase that `node -v` throws an error as it appears that `asdf` is not translating a `.nvmrc` containing `lts/*` to mean Node.js 14.15.4, which is already installed on the system.

> can you verify if you are pointing to the right branch from my repo? I have some work of other pull requests in my master

Me stating "I switched to your branch" was misleading. I manually applied the changes from this PR onto the latest version of `asdf-nodejs`. So, in theory, it should have just this one change to `bin/parse-legacy-file` atop the latest `master` branch of this repository.

> Can you please also send your bash version and OS? It might help with my other work

* bash 3.2.57(1)-release
* macOS 11.1

However, my default shell is zsh, which is version 5.8.

> Also can you verify if is not related to #199?

No, I do not believe it is related to #199. I am not getting any errors similar to what is reported in #199.

---

I will try to find sometime to _actually_ clone your branch to see if the results are the same. Hopefully my responses provide good information in the meantime. Thanks!

augustobmoura commented:
> Me stating "I switched to your branch" was misleading. I manually applied the changes from this PR onto the latest version of `asdf-nodejs`. So, in theory, it should have just this one change to `bin/parse-legacy-file` atop the latest `master` branch of this repository.

Ah ok, applying the changes should also work

It is working as intended, even though the version 14.15.4 is actually the current lts version you need to run `asdf install nodejs lts` (or just `asdf install nodejs` in the directory with the .nvmrc) at least once to link the lts version to the correct number. The error that you got from the last command (the only that would actually link the lts version) is because your bash version is outdated, try upgrading it and repeating the install

> However, my default shell is zsh, which is version 5.8.

Even though your interactive shell is zsh we still use bash as the scripting language for asdf core and asdf-nodejs and recently we added a feature that needs bash version 4+, as mentioned in #199. As I was informed upgrading bash in macOS should be pretty straightforward and recommended for security and features improvements

dcalhoun commented:
@augustobmoura thank you for the detailed guidance. Upgrading `bash` (via `brew install bash`), manually applying the changes from this PR, and running `asdf install nodejs lts` within the directory with the `.nvmrc` does appear to work as expected. 🎉 

Hopefully this PR is merged soon, as it very helpful. Thank you for your work on this!

lokucrazy commented:
@augustobmoura is there anything stopping this from merging?  Would really like to have this feature.

balzdur commented:
@augustobmoura could you please merge this into master ? It would be greatly appreciated.

Stratus3D commented:
@augustobmoura thanks for this PR. What is the status on it? Does this permit indeterminate versions at runtime? I seem to remember we discussed this at one point, but versions in the `.tool-versions` file should be absolute. Looks like these changes pertain to other version files though.

augustobmoura commented:
This PR by itself shouldn't interfere with `.tool-versions` parsing. However, we still have some aliases logic already merged that work with `.tool-versions`. I will work on a new PR to centralize all the logic into the `parse-legacy-file` logic.

I think we can merge this one
